File: kanz-py/kanz_bus/kafka.py
"""Kafka wrapper — connection mgmt + publish/subscribe via aiokafka."""

from __future__ import annotations

import asyncio
from typing import Any

from aiokafka import AIOKafkaConsumer, AIOKafkaProducer

from kanz_bus.bus import Handler, Message


class KafkaClient:
    """Bus client over Kafka (aiokafka).

    Use as an async context manager::

        async with KafkaClient(brokers=["localhost:9092"]) as client:
            await client.publish(Message(subject="market.equity", key=b"AAPL", body=b"..."))
    """

    def __init__(
        self,
        brokers: list[str],
        client_id: str = "kanz",
        publish_timeout: float = 5.0,
    ) -> None:
        if not brokers:
            raise ValueError("kafka: at least one broker required")
        self._brokers = brokers
        self._client_id = client_id
        self._publish_timeout = publish_timeout
        self._producer: AIOKafkaProducer | None = None

    async def connect(self) -> None:
        if self._producer is not None:
            return
        # acks="all" matches the Go writer's RequireAll. Topic provisioning
        # is broker-side (EVT-09: auto-create disabled, topics-job owns it),
        # so we don't pass an auto-create flag.
        self._producer = AIOKafkaProducer(
            bootstrap_servers=",".join(self._brokers),
            client_id=self._client_id,
            acks="all",
            compression_type="snappy",
        )
        await self._producer.start()

    async def close(self) -> None:
        if self._producer is not None:
            await self._producer.stop()
            self._producer = None

    async def __aenter__(self) -> "KafkaClient":
        await self.connect()
        return self

    async def __aexit__(self, *exc: Any) -> None:
        await self.close()

    async def publish(self, msg: Message) -> None:
        if self._producer is None:
            raise RuntimeError("kafka: client not connected; call connect() or use `async with`")
        kafka_headers: list[tuple[str, bytes]] = []
        if msg.headers:
            kafka_headers = [(k, v.encode("utf-8")) for k, v in msg.headers.items()]
        await asyncio.wait_for(
            self._producer.send_and_wait(
                msg.subject,
                value=msg.body,
                key=msg.key,
                headers=kafka_headers,
            ),
            timeout=self._publish_timeout,
        )

    async def subscribe(self, topic: str, group: str, handler: Handler) -> None:
        """Consume ``topic`` under consumer-group ``group``.

        Blocks until the surrounding task is canceled. Handler exceptions
        (other than :class:`asyncio.CancelledError`) skip the commit so the
        message is redelivered on next fetch; normal return commits.
        Bounded retry + DLQ routing layer on top in EVT-18d.
        """
        consumer = AIOKafkaConsumer(
            topic,
            bootstrap_servers=",".join(self._brokers),
            group_id=group,
            client_id=self._client_id,
            enable_auto_commit=False,
        )
        await consumer.start()
        try:
            async for raw in consumer:
                msg = _kafka_to_message(raw)
                try:
                    await handler(msg)
                except Exception:
                    # No commit — message is redelivered on next fetch.
                    continue
                await consumer.commit()
        finally:
            await consumer.stop()


def _kafka_to_message(raw: Any) -> Message:
    headers: dict[str, str] = {}
    if raw.headers:
        for k, v in raw.headers:
            headers[k] = v.decode("utf-8") if v else ""
    return Message(
        subject=raw.topic,
        body=raw.value or b"",
        key=raw.key,
        headers=headers or None,
    )

File: kanz/ts/kanz-bus/src/kafka.ts
/** Kafka wrapper — connection mgmt + publish/subscribe via kafkajs. */

import {
  CompressionCodecs,
  CompressionTypes,
  type Consumer,
  Kafka,
  type KafkaMessage,
  type Producer,
} from "kafkajs";
// @ts-expect-error — kafkajs-snappy ships without bundled types.
import SnappyCodec from "kafkajs-snappy";

import type { Client, Handler, Message } from "./bus.js";

// Snappy must be wired up before the first send/receive that touches a
// snappy-compressed batch. Idempotent — the assignment is harmless if
// repeated across multiple KafkaClient instances. Matches the Go writer's
// kafka.Snappy and the Python aiokafka producer's compression_type="snappy".
CompressionCodecs[CompressionTypes.Snappy] = SnappyCodec;

export interface KafkaConfig {
  brokers: string[];
  clientId?: string;
  /** Per-publish timeout in ms (kafkajs `timeout`). Default 5_000. */
  publishTimeoutMs?: number;
}

/**
 * Bus client over Kafka (kafkajs).
 *
 * ```ts
 * const client = new KafkaClient({ brokers: ["localhost:9092"] });
 * await client.connect();
 * await client.publish({ subject: "market.equity", key: bytes, body: bytes });
 * await client.close();
 * ```
 */
export class KafkaClient implements Client {
  private readonly cfg: Required<KafkaConfig>;
  private readonly kafka: Kafka;
  private producer: Producer | null = null;
  private readonly consumers: Set<Consumer> = new Set();

  constructor(cfg: KafkaConfig) {
    if (!cfg.brokers || cfg.brokers.length === 0) {
      throw new Error("kafka: at least one broker required");
    }
    this.cfg = {
      brokers: cfg.brokers,
      clientId: cfg.clientId ?? "kanz",
      publishTimeoutMs: cfg.publishTimeoutMs ?? 5_000,
    };
    this.kafka = new Kafka({
      brokers: this.cfg.brokers,
      clientId: this.cfg.clientId,
    });
  }

  async connect(): Promise<void> {
    if (this.producer) return;
    // allowAutoTopicCreation matches the Go writer's
    // AllowAutoTopicCreation=false (EVT-09: topics provisioned explicitly
    // by the topics bootstrap Job, never by the client).
    this.producer = this.kafka.producer({ allowAutoTopicCreation: false });
    await this.producer.connect();
  }

  async close(): Promise<void> {
    const errs: unknown[] = [];
    if (this.producer) {
      try {
        await this.producer.disconnect();
      } catch (e) {
        errs.push(e);
      }
      this.producer = null;
    }
    for (const c of this.consumers) {
      try {
        await c.disconnect();
      } catch (e) {
        errs.push(e);
      }
    }
    this.consumers.clear();
    if (errs.length > 0) throw errs[0];
  }

  async publish(msg: Message): Promise<void> {
    if (!this.producer) {
      throw new Error("kafka: client not connected; call connect() first");
    }
    const headers: Record<string, string> = msg.headers ?? {};
    await this.producer.send({
      topic: msg.subject,
      compression: CompressionTypes.Snappy,
      timeout: this.cfg.publishTimeoutMs,
      acks: -1, // RequireAll — matches Go writer
      messages: [
        {
          key: msg.key ? Buffer.from(msg.key) : null,
          value: Buffer.from(msg.body),
          headers,
        },
      ],
    });
  }

  /**
   * Consume `topic` under consumer-group `group`. Resolves when `signal`
   * is aborted. Handler exceptions skip the commit so the message is
   * redelivered after consumer restart / rebalance; normal return
   * commits. Bounded retry + DLQ routing layer on top in a later EVT-19
   * subtask.
   */
  async subscribe(
    topic: string,
    group: string,
    handler: Handler,
    signal?: AbortSignal,
  ): Promise<void> {
    const consumer = this.kafka.consumer({ groupId: group });
    this.consumers.add(consumer);
    await consumer.connect();
    await consumer.subscribe({ topic, fromBeginning: false });
    await consumer.run({
      autoCommit: false,
      eachMessage: async ({ topic: t, partition, message }) => {
        const msg = kafkaToMessage(t, message);
        try {
          await handler(msg);
        } catch {
          // No commit — message is redelivered after restart / rebalance.
          return;
        }
        await consumer.commitOffsets([
          { topic: t, partition, offset: (BigInt(message.offset) + 1n).toString() },
        ]);
      },
    });
    try {
      await waitForAbort(signal);
    } finally {
      try {
        await consumer.disconnect();
      } finally {
        this.consumers.delete(consumer);
      }
    }
  }
}

function kafkaToMessage(topic: string, m: KafkaMessage): Message {
  const out: Message = {
    subject: topic,
    body: m.value ? new Uint8Array(m.value) : new Uint8Array(),
  };
  if (m.key) out.key = new Uint8Array(m.key);
  if (m.headers) {
    const headers: Record<string, string> = {};
    for (const [k, v] of Object.entries(m.headers)) {
      if (v == null) {
        headers[k] = "";
      } else if (typeof v === "string") {
        headers[k] = v;
      } else if (Array.isArray(v)) {
        const first = v[0];
        headers[k] = first == null ? "" : Buffer.from(first).toString("utf-8");
      } else {
        headers[k] = Buffer.from(v).toString("utf-8");
      }
    }
    if (Object.keys(headers).length > 0) out.headers = headers;
  }
  return out;
}

function waitForAbort(signal?: AbortSignal): Promise<void> {
  if (!signal) return new Promise(() => {}); // never resolves
  if (signal.aborted) return Promise.resolve();
  return new Promise((resolve) => {
    signal.addEventListener("abort", () => resolve(), { once: true });
  });
}
